library(biomaRt)
library(siggenes)

library(log4r)
logger <- create.logger()
logfile(logger) <- file.path('../logs/filter.log')
level(logger) <- log4r:::DEBUG
debug(logger,"\n\n--filter log begins--")


filter_probesets <- function(exprset,by){
    exprset[by(exprset),]
}

has_human_homologue <- function(exprset,chip="affy_mogene_1_0_st_v1",
    dataset="mmusculus_gene_ensembl"){
    ensembl_mm <- useMart("ensembl", dataset = dataset)
    mm2hs_filters <- c(chip,"with_hsapiens_homolog")
    mm2hs_gene_atts <- c(chip,"ensembl_gene_id")
    # the names in these lists are arbitrary
    mm2hs_value = list(affyid=rownames(fData(exprset)),with_homolog=TRUE)
    # get the human genes and mouse orthologues
    mm2hs_gene <- getBM(
        attributes = mm2hs_gene_atts ,
        filters = mm2hs_filters,
        value = mm2hs_value, 
        mart = ensembl_mm
    )
    # test
    rownames(fData(exprset))%in%mm2hs_gene[,chip]
}

is_TF <- function(exprset){
    # TODO!
    rownames(exprset)
}

apply_by_phenotype <- function(df,fun,margin=1,phenotype=colnames(df)){
    # this function applies `fun` to the data.frame, grouping by phenotype
    sapply(unique(colnames(df)), function(x) {
        apply(df[,colnames(df) == x],1,fun)
    })
}

strip_unique_identifiers <- function(c){
    # this pulls off the ".x" identifiers generated by make.unique
    sapply(c,function(x){
        strsplit(x,".",fixed=T)[[1]][1]
    })
}

is_differentially_expressed <-function(exprset){
    data <- exprs(exprset)
    colnames(data) <- pData(exprset)$phenotype
    # for every combo of day, find all the differentially expressed genes
    g = c()
    ptypes = unique(pData(exprset)$phenotype)
    debug(logger,"running through phenotypes")
    for (i in seq(length(ptypes)-1)){
        ptype1 = ptypes[i]
        for (ptype2 in ptypes[(i+1):length(ptypes)]){
            cols = (colnames(data) == ptype1) | (colnames(data) == ptype2)
            d = data[,cols]
            cl = strip_unique_identifiers(colnames(d))
            sam.out <- sam(d,cl)
            delta <- findDelta(sam.out,fdr=0.05)
            if (length(delta)>1) delta = delta[2]
            if (!is.null(delta)){
                gi = list.siggenes(sam.out,delta)
                debug(logger,paste("found",length(gi),
                    "differentially expressed genes between the",
                    ptype1,"and",ptype2,"phenotypes"))
                g <- union(g,gi)
            }
        }
    }
    debug(logger,paste("found",
        length(g),"differentially expressed genes in total"))
    rownames(fData(exprset))%in%g 
}
